File: controllers/complaintController.go
package controllers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"os"

	"github.com/gofiber/fiber/v2"
	"rail-madad/config"
	"google.golang.org/grpc/status"
	"google.golang.org/grpc/codes"

)

type PNRRequest struct {
	PNR     string                `json:"pnr"`
	Subject string                `json:"subject"`
	Image   *multipart.FileHeader `form:"image"`
}

const (
	API_URL = "https://api-inference.huggingface.co/models/Salesforce/blip-image-captioning-large"
	API_KEY = "Bearer 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"
)

func queryImageCaption(imageData []byte) (string, error) {
	client := &http.Client{}
	req, err := http.NewRequest("POST", API_URL, bytes.NewBuffer(imageData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", API_KEY)
	req.Header.Set("Content-Type", "application/octet-stream")

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("Hugging Face API error: %s", resp.Status)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	// Log the full response body for debugging
	log.Printf("Hugging Face API Response: %s", body)

	var response []map[string]interface{}
	if err := json.Unmarshal(body, &response); err != nil {
		return "", err
	}

	if len(response) == 0 {
		return "", fmt.Errorf("no response received")
	}

	// Access generated_text instead of caption
	caption, ok := response[0]["generated_text"].(string)
	if !ok {
		return "", fmt.Errorf("invalid response format: %v", response[0])
	}
	return caption, nil
}



func UploadImageToBucket(file *multipart.FileHeader) (string, error) {
	ctx := context.Background()

	bucketName := os.Getenv("GCS_BUCKET_NAME")
	if bucketName == "" {
		return "", fmt.Errorf("bucket name is not configured")
	}

	bucket := config.StorageClient.Bucket(bucketName)

	src, err := file.Open()
	if err != nil {
		return "", err
	}
	defer src.Close()

	data, err := ioutil.ReadAll(src)
	if err != nil {
		return "", err
	}

	obj := bucket.Object("images/" + file.Filename)
	writer := obj.NewWriter(ctx)
	defer writer.Close()

	if _, err := writer.Write(data); err != nil {
		return "", err
	}

	return fmt.Sprintf("https://storage.googleapis.com/%s/images/%s", bucketName, file.Filename), nil
}

func SubmitPNR(c *fiber.Ctx) error {
	var pnrRequest PNRRequest

	pnrRequest.PNR = c.FormValue("pnr")
	pnrRequest.Subject = c.FormValue("subject")
	image, err := c.FormFile("image")
	if err != nil {
		log.Printf("Error getting image: %v", err)
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Failed to get image from request",
		})
	}
	pnrRequest.Image = image

	log.Printf("Received PNR: %s", pnrRequest.PNR)
	log.Println("Image provided in the request")

	imageURL, err := UploadImageToBucket(pnrRequest.Image)
	if err != nil {
		log.Printf("Error uploading image: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to upload image",
		})
	}

	src, err := image.Open()
	if err != nil {
		log.Printf("Error opening image file: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to open image for captioning",
		})
	}
	defer src.Close()

	imageData, err := ioutil.ReadAll(src)
	if err != nil {
		log.Printf("Error reading image file: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to read image for captioning",
		})
	}

	queryGenerated, err := queryImageCaption(imageData)
	if err != nil {
		log.Printf("Error querying image caption: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to generate image caption",
		})
	}

	_, err = config.FirestoreClient.Collection("pnrs").Doc(pnrRequest.PNR).Set(context.Background(), map[string]interface{}{
		"pnr":           pnrRequest.PNR,
		"subject":       pnrRequest.Subject,
		"image":         imageURL,
		"queryGenerated": queryGenerated, 
		"status":        "Pending",
	})
	if err != nil {
		log.Printf("Error storing PNR: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to store PNR",
		})
	}

	return c.JSON(fiber.Map{
		"message":       "PNR submitted successfully",
		"pnr":          pnrRequest.PNR,
		"image":        imageURL,
		"queryGenerated": queryGenerated, 
		"subject":      pnrRequest.Subject,
		"status":       "Pending",
	})
}

func GetComplaints(c *fiber.Ctx) error {
	pnr := c.Params("pnr")
	if pnr == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "PNR is required",
		})
	}

	doc, err := config.FirestoreClient.Collection("pnrs").Doc(pnr).Get(context.Background())
	if err != nil {
		if status.Code(err) == codes.NotFound {
			return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
				"error": "PNR not found",
			})
		}
		log.Printf("Error fetching PNR: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to fetch PNR",
		})
	}

	var pnrData map[string]interface{}
	if err := doc.DataTo(&pnrData); err != nil {
		log.Printf("Error converting PNR data: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to process PNR data",
		})
	}

	return c.JSON(fiber.Map{
		"pnr":           pnr,
		"subject":       pnrData["subject"],
		"queryGenerated": pnrData["queryGenerated"],
		"image":         pnrData["image"],
		"status":        pnrData["status"],
	})
}

File: controllers/complaintController.js
import axios from 'axios';
import admin from '../config/firebase.js';
import dotenv from 'dotenv';

dotenv.config();

const API_URL = process.env.HUGGING_FACE_API_URL;
const API_KEY = process.env.HUGGING_FACE_API_KEY;

const bucketName = process.env.GCS_BUCKET_NAME;

async function queryImageCaption(imageData) {
    try {
        const response = await axios.post(API_URL, imageData, {
            headers: {
                'Authorization': API_KEY,
                'Content-Type': 'application/octet-stream'
            }
        });

        console.log("Hugging Face API Response:", response.data);

        if (response.data && response.data.length > 0 && response.data[0].generated_text) {
            return response.data[0].generated_text;
        } else {
            throw new Error("Invalid response format");
        }
    } catch (error) {
        console.error("Error querying image caption:", error);
        throw error;
    }
}

async function uploadImageToBucket(file) {
    if (!bucketName) {
        throw new Error("Bucket name is not configured");
    }

    const bucket = admin.storage().bucket(bucketName);
    const gcsFileName = `images/${Date.now()}-${file.originalname}`;
    const fileBuffer = file.buffer;

    const bucketFile = bucket.file(gcsFileName);

    try {
        await bucketFile.save(fileBuffer, {
            metadata: {
                contentType: file.mimetype,
            },
        });

        // Generate a download URL that doesn't expire
        const [url] = await bucketFile.getSignedUrl({
            action: 'read',
            expires: '03-01-2500',
        });

        return url;
    } catch (error) {
        console.error("Error uploading to bucket:", error);
        throw error;
    }
}

export const submitPNR = async (req, res) => {
    try {
        const { pnr, subject } = req.body;
        const image = req.file;

        if (!image) {
            return res.status(400).json({ error: "Failed to get image from request" });
        }

        console.log("Received PNR:", pnr);
        console.log("Image provided in the request");

        const imageUrl = await uploadImageToBucket(image);
        const queryGenerated = await queryImageCaption(image.buffer);

        await admin.firestore().collection('pnrs').doc(pnr).set({
            pnr,
            subject,
            imageUrl,  // Store the download URL
            queryGenerated,
            status: "Pending"
        });

        res.json({
            message: "PNR submitted successfully",
            pnr,
            imageUrl,  // Return the download URL in the response
            queryGenerated,
            subject,
            status: "Pending"
        });
    } catch (error) {
        console.error("Error in submit-pnr:", error);
        res.status(500).json({ error: "Internal server error" });
    }
};

export const getComplaints = async (req, res) => {
    try {
        const { pnr } = req.params;

        if (!pnr) {
            return res.status(400).json({ error: "PNR is required" });
        }

        const doc = await admin.firestore().collection('pnrs').doc(pnr).get();

        if (!doc.exists) {
            return res.status(404).json({ error: "PNR not found" });
        }

        const pnrData = doc.data();

        res.json({
            pnr,
            subject: pnrData.subject,
            queryGenerated: pnrData.queryGenerated,
            imageUrl: pnrData.imageUrl, 
            status: pnrData.status
        });
    } catch (error) {
        console.error("Error in get-complaints:", error);
        res.status(500).json({ error: "Internal server error" });
    }
};